src/tof/model.py

```diff
@@ -62,177 +78,187 @@ def run(self, npulses: int = 1):
         initial_mask = sc.ones(
             sizes=self.pulse.birth_times.sizes, unit=None, dtype=bool
         )
-        for comp in components:
+
+        result_choppers = {}
+        result_detectors = {}
+        for c in components:
+            comp = Component(**c.to_dict())
             comp._wavelengths = self.pulse.wavelengths
             t = self.pulse.birth_times + comp.distance / self.pulse.speeds
             comp._arrival_times = t
-            if isinstance(comp, Detector):
+            if isinstance(c, Detector):
                 comp._mask = initial_mask
+                result_detectors[comp.name] = comp
                 continue
             m = sc.zeros(sizes=t.sizes, unit=None, dtype=bool)
-            to = comp.open_times
-            tc = comp.close_times
+            to = c.open_times
+            tc = c.close_times
             for i in range(len(to)):
                 m |= (t > to[i]) & (t < tc[i])
             combined = initial_mask & m
             comp._mask = combined
             comp._own_mask = ~m & initial_mask
             initial_mask = combined
+            result_choppers[comp.name] = comp
 
-    def _add_rays(self, ax, tofs, birth_times, distances, cbar=True, wavelengths=None):
-        x0 = birth_times.to(unit='us', copy=False).values.reshape(-1, 1)
-        x1 = tofs.to(unit='us', copy=False).values.reshape(-1, 1)
-        y0 = np.zeros(x0.size).reshape(-1, 1)
-        y1 = distances.values.reshape(-1, 1)
-        segments = np.concatenate(
-            (
-                np.concatenate((x0, y0), axis=1).reshape(-1, 1, 2),
-                np.concatenate((x1, y1), axis=1).reshape(-1, 1, 2),
-            ),
-            axis=1,
+        return Result(
+            pulse=self.pulse, choppers=result_choppers, detectors=result_detectors
```

Comment:
So these are not instances of `Chopper` and `Detector`?

Reply:
These are now Readonly chopper-like and detector-like objects